File: HW2/practical/2.go
package main

import (
	"fmt"
	"sort"
)

type HopcroftKarp struct {
	graph        map[rune]map[rune]bool
	pairU        map[rune]rune
	pairV        map[rune]rune
	dist         map[rune]int
	maxMatching  int
}

func NewHopcroftKarp(graph map[rune]map[rune]bool) *HopcroftKarp {
	return &HopcroftKarp{
		graph:       graph,
		pairU:       make(map[rune]rune),
		pairV:       make(map[rune]rune),
		dist:        make(map[rune]int),
		maxMatching: 0,
	}
}

func (hk *HopcroftKarp) bfs() bool {
	queue := make([]rune, 0)
	for u := range hk.graph {
		if hk.pairU[u] == 0 {
			hk.dist[u] = 0
			queue = append(queue, u)
		} else {
			hk.dist[u] = -1
		}
	}

	hk.dist[0] = -1
	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]
		if u != 0 {
			for v := range hk.graph[u] {
				if hk.dist[hk.pairV[v]] == -1 {
					hk.dist[hk.pairV[v]] = hk.dist[u] + 1
					queue = append(queue, hk.pairV[v])
				}
			}
		}
	}
	return hk.dist[0] != -1
}

func (hk *HopcroftKarp) dfs(u rune) bool {
	if u != 0 {
		for v := range hk.graph[u] {
			if hk.dist[hk.pairV[v]] == hk.dist[u]+1 {
				if hk.dfs(hk.pairV[v]) {
					hk.pairV[v] = u
					hk.pairU[u] = v
					return true
				}
			}
		}
		hk.dist[u] = -1
		return false
	}
	return true
}

func (hk *HopcroftKarp) maxMatchingSize() int {
	for hk.bfs() {
		for u := range hk.graph {
			if hk.pairU[u] == 0 && hk.dfs(u) {
				hk.maxMatching++
			}
		}
	}
	return hk.maxMatching
}

func parseInput(inputStr string) (map[rune]map[rune]bool, map[rune]bool, map[string]bool) {
	graph := make(map[rune]map[rune]bool)
	vertices := make(map[rune]bool)
	edges := make(map[string]bool)

	i := 0
	sequence := []rune(inputStr)
	for i < len(sequence) {
		if sequence[i] == ' ' {
			i++
			continue
		} else if (i == len(sequence)-1) || sequence[i+1] == ' ' {
			isolatedVertex := sequence[i]
			vertices[isolatedVertex] = true
			if _, ok := graph[isolatedVertex]; !ok {
				graph[isolatedVertex] = make(map[rune]bool)
			}
			i++
			continue
		}
		vertex1 := sequence[i]
		vertex2 := sequence[i+1]
		edge := string(sortRunes([]rune{vertex1, vertex2}))
		vertices[vertex1] = true
		vertices[vertex2] = true
		edges[edge] = true
		if _, ok := graph[vertex1]; !ok {
			graph[vertex1] = make(map[rune]bool)
		}
		if _, ok := graph[vertex2]; !ok {
			graph[vertex2] = make(map[rune]bool)
		}
		graph[vertex1][vertex2] = true
		graph[vertex2][vertex1] = true
		i += 2
	}
	return graph, vertices, edges
}

func maximumMatching(inputString string) int {
	graph, _, _ := parseInput(inputString)
	hk := NewHopcroftKarp(graph)
	return hk.maxMatchingSize()
}

func sortRunes(runes []rune) []rune {
	sort.Slice(runes, func(i, j int) bool { return runes[i] < runes[j] })
	return runes
}

func main() {
	fmt.Println(maximumMatching("abacdedfsopwkslskc")) // Output: 5
	fmt.Println(maximumMatching("abacbcbebddeegeffgghhigiijkjkljllnlmmnnpnopororqoqqtsqstsuuvuwvwwxwyxyzx")) // Output: 13
}

File: HW2/practical/2.rs
use std::collections::{HashMap, HashSet};

struct HopcroftKarp<'a> {
    graph: &'a HashMap<char, HashSet<char>>,
    pair_u: HashMap<char, Option<char>>,
    pair_v: HashMap<char, Option<char>>,
    dist: HashMap<Option<char>, usize>,
    max_matching: usize,
}

impl<'a> HopcroftKarp<'a> {
    fn new(graph: &'a HashMap<char, HashSet<char>>) -> Self {
        let mut pair_u = HashMap::new();
        let mut pair_v = HashMap::new();
        for (u, _) in graph.iter() {
            pair_u.insert(*u, None);
        }
        for (v, _) in graph.values().flatten() {
            pair_v.insert(*v, None);
        }
        let mut dist = HashMap::new();
        dist.insert(None, usize::MAX);
        for u in graph.keys() {
            if pair_u[u].is_none() {
                dist.insert(Some(*u), 0);
            } else {
                dist.insert(Some(*u), usize::MAX);
            }
        }
        HopcroftKarp {
            graph,
            pair_u,
            pair_v,
            dist,
            max_matching: 0,
        }
    }

    fn bfs(&mut self) -> bool {
        let mut queue = Vec::new();
        for u in self.graph.keys() {
            if self.pair_u[u].is_none() {
                self.dist.insert(Some(*u), 0);
                queue.push(*u);
            } else {
                self.dist.insert(Some(*u), usize::MAX);
            }
        }
        self.dist.insert(None, usize::MAX);

        while let Some(u) = queue.pop() {
            if let Some(d) = self.dist.get(&Some(u)) {
                for &v in self.graph[&u].iter() {
                    if *self.dist.get(&self.pair_v[&v]).unwrap() == usize::MAX {
                        *self.dist.get_mut(&self.pair_v[&v]).unwrap() = *d + 1;
                        queue.push(self.pair_v[&v].unwrap());
                    }
                }
            }
        }
        *self.dist.get(&None).unwrap() != usize::MAX
    }

    fn dfs(&mut self, u: char) -> bool {
        if let Some(d) = self.dist.get(&Some(u)) {
            for &v in self.graph[&u].iter() {
                if *self.dist.get(&self.pair_v[&v]).unwrap() == *d + 1 {
                    if self.pair_v[&v].is_none() || self.dfs(self.pair_v[&v].unwrap()) {
                        self.pair_v.insert(v, Some(u));
                        self.pair_u.insert(u, Some(v));
                        return true;
                    }
                }
            }
            self.dist.insert(Some(u), usize::MAX);
            return false;
        }
        true
    }

    fn max_matching_size(&mut self) -> usize {
        while self.bfs() {
            for &u in self.graph.keys() {
                if self.pair_u[&u].is_none() && self.dfs(u) {
                    self.max_matching += 1;
                }
            }
        }
        self.max_matching / 2
    }
}

fn parse_input(input_str: &str) -> (HashMap<char, HashSet<char>>, HashSet<char>, HashSet<String>) {
    let mut graph = HashMap::new();
    let mut vertices = HashSet::new();
    let mut edges = HashSet::new();

    let mut i = 0;
    let sequence = input_str.chars().collect::<Vec<char>>();
    while i < sequence.len() {
        if sequence[i] == ' ' {
            i += 1;
            continue;
        } else if i == sequence.len() - 1 || sequence[i + 1] == ' ' {
            let isolated_vertex = sequence[i];
            vertices.insert(isolated_vertex);
            graph.entry(isolated_vertex).or_insert(HashSet::new());
            i += 1;
            continue;
        }
        let vertex1 = sequence[i];
        let vertex2 = sequence[i + 1];
        let mut edge_chars = vec![vertex1, vertex2];
        edge_chars.sort();
        let edge = edge_chars.into_iter().collect::<String>();
        vertices.insert(vertex1);
        vertices.insert(vertex2);
        edges.insert(edge.clone());
        graph.entry(vertex1).or_insert(HashSet::new()).insert(vertex2);
        graph.entry(vertex2).or_insert(HashSet::new()).insert(vertex1);
        i += 2;
    }
    (graph, vertices, edges)
}

fn maximum_matching(input_string: &str) -> usize {
    let (graph, _, _) = parse_input(input_string);
    let mut hk = HopcroftKarp::new(&graph);
    hk.max_matching_size()
}

fn main() {
    let input_str = {
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).expect("Failed to read line");
        input.trim().to_string()
    };
    println!("{}", maximum_matching(&input_str));
}
